feat: incorporate multi-file composition with ci docker test environments

bradymiller opened the pull request:
fixes #8426

This is more of a "see what happens" PR. Just seeing if using include in the docker-compose stuff for ci is possible. Good chance I have done this all wrong and all testing will fail :)

bradymiller commented:
Turn out include mechanism can not have duplicate service names, so fare thee well to this approach :)

bradymiller commented:
Ended up changing strategy from using include directives to multi-file composition (compose file merging). Adds a bit of complexity since needs to be run from command line, however makes the end result docker-compose.yml simpler when adding or changing test configurations.

bradymiller commented:
hi @kojiromike . Let me know your thoughts on this strategy. I never would of been able to do this with your prior centralization work. I tried the include directives, but that failed since not allowed to modify services in separate includes. The command line docker compose merging is much more flexible and allows the merging of service changes etc. It does admittedly complicate things a bit since can no longer just do something to the docker-compose.yml without adding the other 2 along with the f parameters. Also the shares yml are a bit fragmented since only included needed elements (can use the config command on the merge to see what the end script actually looks like, though). Goal was to make the most recent thing we do (add/modify testing environments) as simple as possible.

kojiromike reviewed (approved):
This is fantastic. It'll make it much easier to add new test configurations on in the future.

bradymiller commented on .github/workflows/test.yml:
```diff
@@ -35,6 +35,10 @@ jobs:
 
+            # collect docker-compose.yml templates that will be used in the docker compose up merge
+            webserver_template=$(yq '.x-includes.webserver-template' "ci/${docker_dir}/docker-compose.yml")
+            database_template=$(yq '.x-includes.database-template' "ci/${docker_dir}/docker-compose.yml")
```
very nice suggestion. incorporated COMPOSE_FILE.

bradymiller commented on ci/ciLibrary.source:
```diff
@@ -48,7 +48,7 @@ composer_github_auth() {
 # but it's rarely used in the same context as docker compose.
 dc() {
-    docker compose --project-directory "ci/${DOCKER_DIR}" "$@"
+    docker compose -f "ci/${WEBSERVER_TEMPLATE}" -f "ci/${DATABASE_TEMPLATE}" -f "ci/${DOCKER_DIR}/docker-compose.yml" "$@"
```
done, gave it an official step

bradymiller commented on ci/compose-shared-nginx.yml:
```diff
@@ -1,20 +1,16 @@
-      service: mysql
+    restart: always
+    environment:
+      MYSQL_ROOT_PASSWORD: root
```
agreed and put it back in its rightful place

bradymiller commented on ci/compose-shared-apache.yml:
```diff
@@ -0,0 +1,15 @@
+services:
+  mysql:
+    restart: always
```
just habit. removed them.

bradymiller commented:
hi @stephenwaite and @kojiromike , Thanks for the suggestions. Fixed documentation typo and added couple more debugging examples (including config). Bringing in.
